### PR TITLE
Top Posts Widget: add a new option to choose what Post Types to display

### PR DESCRIPTION
**Pull request by jeherve**
Another attempt at fixing #54 and #943

I would appreciate a review of my changes, especially the data validation part, as I've struggled a bit with it.

Thanks!

### CODE REVIEW

File: modules/widgets/top-posts.php [hunk 1]
```diff
@@ -105,6 +130,14 @@ function update( $new_instance, $old_instance ) {
 			$instance['count'] = 10;
 		}
 
+		$allowed_post_types = array_values( get_post_types( array( 'public' => true ) ) );
+		$instance['types'] = $new_instance['types'];
+		foreach( $new_instance['types'] as $type ) {
+			if ( ! in_array( $type, $allowed_post_types ) ) {
+				unset( $type );
```
Comment (zinigor) [inline modules/widgets/top-posts.php]: This won't actually unset, you need to unset a member of the `$new_isntance['types']` array by key.

File: modules/widgets/top-posts.php [hunk 2]
```diff
@@ -143,12 +187,31 @@ function widget( $args, $instance ) {
...
+			 */
 			$get_image_options = apply_filters( 'jetpack_top_posts_widget_image_options', $get_image_options );
 		}
 
 		$posts = $this->get_by_views( $count );
 
-		if ( !$posts ) {
+		// Filter the returned posts. Remove all posts that do not match the chosen Post Types.
+		foreach ( $posts as $k => $post ) {
+			if ( isset( $types ) && ! in_array( $post['post_type'], $types ) ) {
```
Comment (zinigor) [inline modules/widgets/top-posts.php]: If `$types` is not set, then this whole cycle is pointless. The check should be done outside the cycle, don't you think?

**Comment by zinigor** [comment]
So far looks nice, let's discuss my comments when you have a moment. Also, I would suggest doing separate PRs for doc changes and code logic changes, but in this case it should be fine since it's only one file.

File: modules/widgets/top-posts.php (hunk 2 above)
Comment (jeherve) [inline modules/widgets/top-posts.php]: That makes sense. Fixed in 54f5465441d6669787965d5ef6de7467eeeacb24

File: modules/widgets/top-posts.php (hunk 1 above)
Comment (jeherve) [inline modules/widgets/top-posts.php]: I'm not sure I understand, don't we want to remove the whole `$type` if it is not allowed?

Comment (zinigor) [inline modules/widgets/top-posts.php]: No, I mean if you `unset( $type )`, it won't actually unset the array element, the `$new_instance['types'][ $whatever_the_key_for_that_type ]` will still remain in the array, you need to `unset( $new_instance['types'][ $whatever_the_key_for_that_type ] )` for it to work the way you intended.

Comment (jeherve) [inline modules/widgets/top-posts.php]: Got it. Thanks for the clarification in Slack! I made some changes in 8de214b2c6bd5ef0a46c1c0090be8acc8ed4d6bd. Tested them, everything seems to work.

**Comment by dereksmart** [comment]
read through and tested.  Nice work @jeherve

**Pull request merged**